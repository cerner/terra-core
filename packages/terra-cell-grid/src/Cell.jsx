import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Cell.module.scss';
import { styleFromWidth } from './_CellGridUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Content to be displayed for the cell.
   */
  children: PropTypes.node,
  /**
   * Function callback to get the ref of the cell.
   */
  refCallback: PropTypes.func,
  /**
   * Composed width for the cell. Can be provided as either a static, percentage, or scalar value.
   */
  width: PropTypes.shape({
    static: PropTypes.shape({
      value: PropTypes.number.isRequired,
      unit: PropTypes.string.isRequired,
    }),
    percentage: PropTypes.number,
    scalar: PropTypes.number,
  }),
};

const defaultProps = {
  children: [],
};

const Cell = ({
  children,
  refCallback,
  width,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
  ]);

  return (
    <div
      {...customProps}
      style={styleFromWidth(width)} // eslint-disable-line react/forbid-dom-props
      className={customProps.className ? `${cellClassNames} ${customProps.className}` : cellClassNames}
      ref={refCallback}
    >
      {children}
    </div>
  );
};

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default Cell;
