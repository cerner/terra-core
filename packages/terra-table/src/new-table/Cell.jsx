import React, {
  useContext, useState, useRef, useEffect,
} from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import * as KeyCode from 'keycode-js';
import classNames from 'classnames/bind';
import VisuallyHiddenText from 'terra-visually-hidden-text';
import ThemeContext from 'terra-theme-context';
import FocusTrap from 'focus-trap-react';
import styles from './Cell.module.scss';
import ColumnContext from './utils/ColumnContext';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifier of the row in which the Cell will be rendered.
   */
  rowId: PropTypes.string.isRequired,

  /**
   * String identifier of the column in which the Cell will be rendered.
   */
  columnId: PropTypes.string.isRequired,

  /**
   * The cell's row position in the grid. This is zero based.
   */
  rowIndex: PropTypes.number,

  /**
   * The cell's column position in the grid. This is zero based.
   */
  columnIndex: PropTypes.number,

  /**
   * Content that will be rendered within the Cell.
   */
  children: PropTypes.node,

  /**
   *  Boolean indicating if cell contents are masked.
   */
  isMasked: PropTypes.bool,

  /**
   * Boolean value indicating whether or not the column header is selectable.
   */
  isSelectable: PropTypes.bool,

  /**
   * Boolean indicating whether the Cell is currently selected.
   */
  isSelected: PropTypes.bool,

  /**
   * String that labels the cell for accessibility.
   */
  ariaLabel: PropTypes.string,

  /**
   * Boolean indicating that the cell is a row header.
   */
  isRowHeader: PropTypes.bool,

  /**
   * Boolean indicating that the cell has been highlighted.
   */
  isHighlighted: PropTypes.bool,

  /**
   * Callback function that will be called when this cell is selected.
   */
  onCellSelect: PropTypes.func,

  /**
   * String that specifies the height of the cell. Any valid CSS value is accepted.
   */
  height: PropTypes.string,

  /**
   * @private
   * The intl object containing translations. This is retrieved from the context automatically by injectIntl.
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
};

const defaultProps = {
  isRowHeader: false,
  isSelected: false,
  isSelectable: true,
  isMasked: false,
};

function Cell(props) {
  const {
    rowId,
    columnId,
    rowIndex,
    columnIndex,
    ariaLabel,
    isMasked,
    isRowHeader,
    isSelectable,
    isSelected,
    isHighlighted,
    children,
    onCellSelect,
    height,
    intl,
  } = props;

  const cellRef = useRef();
  const [isFocusTrapEnabled, setFocusTrapEnabled] = useState(false);
  const [isInteractable, setInteractable] = useState(false);
  const theme = useContext(ThemeContext);
  const columnContext = useContext(ColumnContext);

  /**
   * Determine if cell has focusable elements
   */
  const hasFocusableElements = () => {
    const focusableElementSelector = "a[href]:not([tabindex='-1']), area[href]:not([tabindex='-1']), input:not([disabled]):not([tabindex='-1']), "
    + "select:not([disabled]):not([tabindex='-1']), textarea:not([disabled]):not([tabindex='-1']), button:not([disabled]):not([tabindex='-1']), "
    + "iframe:not([tabindex='-1']), [tabindex]:not([tabindex='-1']), [contentEditable=true]:not([tabindex='-1'])";

    const focusableElements = [...cellRef.current.querySelectorAll(`${focusableElementSelector}`)].filter(el => !el.hasAttribute('disabled') && !el.getAttribute('aria-hidden'));

    return focusableElements.length > 0;
  };

  useEffect(() => {
    setInteractable(hasFocusableElements());
  }, []);

  /**
   * Handles the onDeactivate callback for FocusTrap component
   */
  const deactiveFocusTrap = () => {
    setFocusTrapEnabled(false);
    columnContext.setCellAriaLiveMessage(intl.formatMessage({ id: 'Terra.dataGrid.resume-navigation' }));
  };

  /**
   * Handles mouse down event for cell
   */
  const onMouseDown = ((event) => {
    if (!isFocusTrapEnabled) {
      onCellSelect({
        rowId, columnId, rowIndex, columnIndex, isShiftPressed: event.shiftKey, isCellSelectable: (!isMasked && isSelectable),
      });
    }
  });

  /**
   * Keyboard event handler
   */
  const handleKeyDown = (event) => {
    const key = event.keyCode;

    if (isFocusTrapEnabled) {
      switch (key) {
        case KeyCode.KEY_ESCAPE:
          deactiveFocusTrap();
          break;
        default:
      }

      event.stopPropagation();
    } else {
      switch (key) {
        case KeyCode.KEY_RETURN:
          // Lock focus into component
          if (hasFocusableElements()) {
            setFocusTrapEnabled(true);
            columnContext.setCellAriaLiveMessage(intl.formatMessage({ id: 'Terra.dataGrid.cell-focus-trapped' }));
            event.stopPropagation();
            event.preventDefault();
          }
          break;
        case KeyCode.KEY_SPACE:
          if (onCellSelect) {
            onCellSelect({
              rowId, columnId, rowIndex, columnIndex, isShiftPressed: event.shiftKey, isCellSelectable: (!isMasked && isSelectable),
            });
          }
          event.preventDefault(); // prevent the default scrolling
          break;
        default:
      }
    }
  };

  // Create cell content for masked and blank cells
  let cellContent;
  if (isMasked) {
    cellContent = (
      <span className={cx('no-data-cell', theme.className)}>
        {intl.formatMessage({ id: 'Terra.dataGrid.maskedCell' })}
      </span>
    );
  } else if (!children) {
    cellContent = (
      <span className={cx('no-data-cell', theme.className)}>
        {intl.formatMessage({ id: 'Terra.dataGrid.blank' })}
      </span>
    );
  } else {
    cellContent = children;
  }

  const className = cx('cell', {
    masked: isMasked,
    pinned: columnIndex < columnContext.pinnedColumnOffsets.length,
    selectable: isSelectable && !isMasked,
    selected: isSelected && !isMasked,
    highlighted: isHighlighted,
    blank: !children,
  }, theme.className);

  const cellLeftEdge = (columnIndex < columnContext.pinnedColumnOffsets.length) ? columnContext.pinnedColumnOffsets[columnIndex] : null;

  const CellTag = isRowHeader ? 'th' : 'td';
  return (
    <CellTag
      ref={cellRef}
      aria-selected={isSelected}
      aria-label={ariaLabel}
      tabIndex={-1}
      className={className}
      {...(isRowHeader && { scope: 'row', role: 'rowheader' })}
      onMouseDown={onCellSelect ? onMouseDown : undefined}
      onKeyDown={handleKeyDown}
      style={{ left: cellLeftEdge }} // eslint-disable-line react/forbid-component-props
    >
      {/* Wrap cell content with focus trap */}
      <FocusTrap
        active={isFocusTrapEnabled}
        focusTrapOptions={{
          returnFocusOnDeactivate: true, clickOutsideDeactivates: true, escapeDeactivates: false, onDeactivate: deactiveFocusTrap,
        }}
      >
        {/* eslint-disable-next-line react/forbid-dom-props */}
        <div className={cx('cell-content', theme.className)} style={{ height }}>{cellContent}</div>
      </FocusTrap>
      {isInteractable && <VisuallyHiddenText text={intl.formatMessage({ id: 'Terra.dataGrid.cell-interactable' })} />}
    </CellTag>
  );
}

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default React.memo(injectIntl(Cell));
