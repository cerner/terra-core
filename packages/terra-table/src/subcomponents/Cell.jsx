import React, {
  useContext, useRef,
} from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import classNames from 'classnames/bind';
import VisuallyHiddenText from 'terra-visually-hidden-text';
import ThemeContext from 'terra-theme-context';
import styles from './Cell.module.scss';
import ColumnContext from '../utils/ColumnContext';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifier of the row in which the Cell will be rendered.
   */
  rowId: PropTypes.string.isRequired,

  /**
   * String identifier of the column in which the Cell will be rendered.
   */
  columnId: PropTypes.string.isRequired,

  /**
   * The cell's row position in the grid. This is zero based.
   */
  rowIndex: PropTypes.number,

  /**
   * The cell's column position in the grid. This is zero based.
   */
  columnIndex: PropTypes.number,

  /**
   * Content that will be rendered within the Cell.
   */
  children: PropTypes.node,

  /**
   *  Boolean indicating if cell contents are masked.
   */
  isMasked: PropTypes.bool,

  /**
   * String that labels the cell for accessibility.
   */
  ariaLabel: PropTypes.string,

  /**
   * Boolean indicating that the cell is a row header.
   */
  isRowHeader: PropTypes.bool,

  /**
   * String that specifies the height of the cell. Any valid CSS value is accepted.
   */
  height: PropTypes.string,

  /**
   * @private
   * The intl object containing translations. This is retrieved from the context automatically by injectIntl.
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
};

const defaultProps = {
  isRowHeader: false,
  isMasked: false,
};

function Cell(props) {
  const {
    rowId,
    columnId,
    rowIndex,
    columnIndex,
    ariaLabel,
    isMasked,
    isRowHeader,
    children,
    height,
    intl,
  } = props;

  const cellRef = useRef();
  const theme = useContext(ThemeContext);
  const columnContext = useContext(ColumnContext);

  // Create cell content for masked and blank cells
  let cellContent;
  if (isMasked) {
    cellContent = (
      <span className={cx('no-data-cell', theme.className)}>
        {intl.formatMessage({ id: 'Terra.dataGrid.maskedCell' })}
      </span>
    );
  } else if (!children) {
    cellContent = (
      <span className={cx('no-data-cell', theme.className)}>
        {intl.formatMessage({ id: 'Terra.dataGrid.blank' })}
      </span>
    );
  } else {
    cellContent = children;
  }

  const className = cx('cell', {
    masked: isMasked,
    pinned: columnIndex < columnContext.pinnedColumnOffsets.length,
    blank: !children,
  }, theme.className);

  const cellLeftEdge = (columnIndex < columnContext.pinnedColumnOffsets.length) ? columnContext.pinnedColumnOffsets[columnIndex] : null;

  const CellTag = isRowHeader ? 'th' : 'td';
  return (
    <CellTag
      ref={cellRef}
      aria-label={ariaLabel}
      tabIndex={-1}
      className={className}
      {...(isRowHeader && { scope: 'row', role: 'rowheader' })}
      // eslint-disable-next-line react/forbid-component-props
      style={{ left: cellLeftEdge }}
    >
      {/* eslint-disable-next-line react/forbid-dom-props */}
      <div className={cx('cell-content', theme.className)} style={{ height }}>{cellContent}</div>
      <VisuallyHiddenText text={intl.formatMessage({ id: 'Terra.dataGrid.cell-interactable' })} />
    </CellTag>
  );
}

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default React.memo(injectIntl(Cell));
