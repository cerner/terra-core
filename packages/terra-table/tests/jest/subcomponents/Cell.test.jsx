import React from 'react';
import Cell from '../../../src/subcomponents/_Cell';

describe('Cell', () => {
  // Snapshot Tests
  it('should render a Cell', () => {
    const shallowComponent = shallow(<Cell />);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with child content', () => {
    const shallowComponent = shallow(<Cell>test text</Cell>);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with child content and remove inner', () => {
    const shallowComponent = shallow(<Cell removeInner>test text</Cell>);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with disclosure label', () => {
    const shallowComponent = shallow(<Cell disclosure={{ label: 'test link label' }} />);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with disclosure label and isCurrent', () => {
    const shallowComponent = shallow(<Cell disclosure={{ label: 'test link label', isCurrent: true }} />);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render with callback functions', () => {
    const shallowComponent = shallow(<Cell refCallback={jest.fn()} />);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with width inner container removed', () => {
    const shallowComponent = shallow(<Cell removeInner />);
    expect(shallowComponent).toMatchSnapshot();
  });

  it('should render a Cell with width set', () => {
    const shallowComponent = shallow(<Cell width={{ scalar: 2 }} />);
    expect(shallowComponent).toMatchSnapshot();
  });
});
