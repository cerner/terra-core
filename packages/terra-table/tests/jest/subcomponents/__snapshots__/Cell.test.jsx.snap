// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render a Cell 1`] = `
<div
  className="cell"
  role="gridcell"
>
  <div
    className="container"
  />
</div>
`;

exports[`Cell should render a Cell with child content 1`] = `
<div
  className="cell"
  role="gridcell"
>
  <div
    className="container"
  >
    test text
  </div>
</div>
`;

exports[`Cell should render a Cell with child content and remove inner 1`] = `
<div
  className="cell"
  role="gridcell"
>
  test text
</div>
`;

exports[`Cell should render a Cell with disclosure label 1`] = `
<div
  className="cell"
  disclosure={
    Object {
      "label": "test link label",
    }
  }
  role="gridcell"
>
  <div
    className="container"
  />
</div>
`;

exports[`Cell should render a Cell with disclosure label and isCurrent 1`] = `
<div
  className="cell"
  disclosure={
    Object {
      "isCurrent": true,
      "label": "test link label",
    }
  }
  role="gridcell"
>
  <div
    className="container"
  />
</div>
`;

exports[`Cell should render a Cell with width inner container removed 1`] = `
<div
  className="cell"
  role="gridcell"
/>
`;

exports[`Cell should render a Cell with width set 1`] = `
<div
  className="cell"
  role="gridcell"
  style={
    Object {
      "flex": "2 2 0px",
      "msFlex": "2 2 0px",
    }
  }
>
  <div
    className="container"
  />
</div>
`;

exports[`Cell should render with callback functions 1`] = `
<div
  className="cell"
  role="gridcell"
>
  <div
    className="container"
  />
</div>
`;
